[pt2][cache] rework cache for true generic usage + better tests

nmacchioni opened the pull request:
Differential Revision: D82933509

over the weekend I realized that some of the cache implementation was a bit silly, and too constrained to be actually generic. for example, InMemoryCache[str, bytes] was odd since we'd probably want to be able to store more than just str keys with bytes values. so tldr; everything is now generic, with the one constraint being that Key and Value must both be pickle-able types. this makes things a lot simpler for us, since all caches can now be str -> bytes caches under the hood if we'd like, and Key/Value just get pickled on the way in and out.

with this change, there were also some improvements made to the testing; mainly better coverage, but now we also test each cache across every combination of Key/Value types to ensure that they will work with the types we might specify later

I also hardened some things here and there, for example we now use literal_eval (forgot who mentioned this on the first PR, but thank you for the suggestion!), and all errors coming from the caching will be wrapped in CacheError from now on (although we still raise from the original error context where possible)

putting this PR up now for feedback, in the process of generalizing the code I did remove the documentation since it was becoming outdated but I will add that back in after the PR is green

I have the next PR ready as well (implements a fresh cache context manager), will export once this lands

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

aorenste reviewed (approved):
approving with some minor comments

masnesral commented on torch/_inductor/cache.py:
```diff
@@ -0,0 +1,229 @@
+
+    def _fpath_from_key(self: Self, key: Key) -> Path:
+        try:
+            return self.base_dir / sha256(pickle.dumps(key)).hexdigest()[:32]
```
FYI in codecache.py I think we do `base64.b32encode(hashlib.sha256(data).digest())[:51].decode("utf-8").lower()` for this operation in case that's somehow better. (I know because I refactored this code once, but I didn't dig into why we do it this way)

nmacchioni commented on test/inductor/test_cache.py:
```diff
@@ -0,0 +1,737 @@
+        return {icache.Cache, icache.AsyncCache}
+
+    @staticmethod
+    def cache_types() -> list[type[icache.Cache]]:
```
Sorry, not sure the question here

nmacchioni commented on torch/_inductor/cache.py:
```diff
@@ -0,0 +1,229 @@
+
+        try:
+            with open(fpath, "rb") as fp:
+                cache._cache = pickle.load(fp)
```
So what happens if we update the version of the cache, but the files still exist? get should obviously return None, since the values stored by the previous cache version can't be safely parsed, but should we also remove that file so that the key can be re-cached?

nmacchioni replied:
I think in this case if there is overlap I'll just prefer to delete the old version cache, if the user wants to fiddle with the version then they can be ping ponged all they want haha

nmacchioni committed:
[pt2][cache] rework cache for true generic usage + better tests (#163488)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/163488

Test Plan:
```
buck test fbcode//mode/opt caffe2/test/inductor:cache
```

Differential Revision: D82933509
